File: flutter/lib/services/webrtc_service.dart
import 'dart:convert';
import 'package:flutter_webrtc/flutter_webrtc.dart';
import 'package:socket_io_client/socket_io_client.dart' as io;

/// Callback for WebRTC messages received via data channel
typedef WebRTCMessageHandler = void Function(Map<String, dynamic> data);

/// Callback for connection state changes
typedef ConnectionStateHandler = void Function(String state);

/// WebRTC P2P connection service
class WebRTCService {
  RTCPeerConnection? _peerConnection;
  RTCDataChannel? _dataChannel;
  final io.Socket _socket;
  bool _isConnected = false;
  WebRTCMessageHandler? _messageHandler;
  ConnectionStateHandler? _onConnectionStateChange;

  WebRTCService(this._socket) {
    _setupSignalingHandlers();
  }

  /// Setup signaling event handlers
  void _setupSignalingHandlers() {
    // Handle WebRTC offer from Mac
    _socket.on('webrtc:offer', (data) async {
      print('📡 Received WebRTC offer from Mac');
      final offer = data['offer'] as Map<String, dynamic>;
      await _handleOffer(offer);
    });

    // Handle ICE candidates from Mac
    _socket.on('webrtc:ice-candidate', (data) async {
      print('🧊 Received ICE candidate from Mac');
      final candidate = data['candidate'] as Map<String, dynamic>;
      await _handleIceCandidate(candidate);
    });
  }

  /// Handle incoming WebRTC offer
  Future<void> _handleOffer(Map<String, dynamic> offer) async {
    try {
      // Create peer connection if it doesn't exist
      if (_peerConnection == null) {
        await _createPeerConnection();
      }

      // Set remote description (the offer from Mac)
      final description = RTCSessionDescription(
        offer['sdp'] as String,
        offer['type'] as String,
      );
      await _peerConnection!.setRemoteDescription(description);
      print('✅ Remote description set');

      // Create answer
      final answer = await _peerConnection!.createAnswer();

      // Set local description once
      await _peerConnection!.setLocalDescription(answer);

      // Wait for ICE gathering to complete (with timeout)
      await _waitForIceGathering();

      print('📡 Sending WebRTC answer to Mac (with all ICE candidates)');

      // Send answer to Mac via signaling server
      final localDescription = await _peerConnection!.getLocalDescription();
      _socket.emit('webrtc:answer', {
        'answer': {
          'sdp': localDescription?.sdp,
          'type': localDescription?.type,
        },
      });
    } catch (error) {
      print('❌ Failed to handle offer: $error');
    }
  }

  /// Wait for ICE gathering to complete with timeout
  Future<void> _waitForIceGathering() async {
    if (_peerConnection?.iceGatheringState ==
        RTCIceGatheringState.RTCIceGatheringStateComplete) {
      return;
    }

    await Future.any([
      Future(() async {
        while (_peerConnection?.iceGatheringState !=
            RTCIceGatheringState.RTCIceGatheringStateComplete) {
          await Future.delayed(const Duration(milliseconds: 100));
          print(
              'ICE Gathering State: ${_peerConnection?.iceGatheringState}');
        }
      }),
      Future.delayed(const Duration(seconds: 3), () {
        print('⏱️  ICE gathering timeout - proceeding with available candidates');
      }),
    ]);
  }

  /// Handle incoming ICE candidate
  Future<void> _handleIceCandidate(Map<String, dynamic> candidate) async {
    try {
      if (_peerConnection != null && candidate['candidate'] != null) {
        final iceCandidate = RTCIceCandidate(
          candidate['candidate'] as String,
          candidate['sdpMid'] as String?,
          candidate['sdpMLineIndex'] as int?,
        );
        await _peerConnection!.addCandidate(iceCandidate);
        print('✅ ICE candidate added');
      }
    } catch (error) {
      print('❌ Failed to add ICE candidate: $error');
    }
  }

  /// Create the peer connection
  Future<void> _createPeerConnection() async {
    print('🔗 Creating WebRTC peer connection...');

    // Configuration with STUN server
    final configuration = {
      'iceServers': [
        {'urls': 'stun:stun.l.google.com:19302'},
      ],
      'iceTransportPolicy': 'all',
      'iceCandidatePoolSize': 10,
    };

    _peerConnection = await createPeerConnection(configuration);

    // Handle ICE candidates
    _peerConnection!.onIceCandidate = (RTCIceCandidate candidate) {
      print('🧊 Generated ICE candidate, sending to Mac');
      final candidateStr = candidate.candidate ?? '';
      final parts = candidateStr.split(' ');
      final candidateType = parts.length > 7 ? parts[7] : 'unknown';
      print('   Type: $candidateType, Candidate: ${candidateStr.substring(0, candidateStr.length > 50 ? 50 : candidateStr.length)}...');

      _socket.emit('webrtc:ice-candidate', {
        'candidate': {
          'candidate': candidate.candidate,
          'sdpMid': candidate.sdpMid,
          'sdpMLineIndex': candidate.sdpMLineIndex,
        },
      });
    };

    // Handle connection state changes
    _peerConnection!.onConnectionState = (RTCPeerConnectionState state) {
      final stateStr = _connectionStateToString(state);
      print('WebRTC Connection State: $stateStr');

      if (state == RTCPeerConnectionState.RTCPeerConnectionStateConnected) {
        _isConnected = true;
        print('🎉 WebRTC P2P connection established!');
      } else if (state ==
              RTCPeerConnectionState.RTCPeerConnectionStateDisconnected ||
          state == RTCPeerConnectionState.RTCPeerConnectionStateFailed ||
          state == RTCPeerConnectionState.RTCPeerConnectionStateClosed) {
        _isConnected = false;
        print('❌ WebRTC connection lost');
      }

      _onConnectionStateChange?.call(stateStr);
    };

    // Handle ICE connection state changes
    _peerConnection!.onIceConnectionState = (RTCIceConnectionState state) {
      print('ICE Connection State: $state');
      if (state == RTCIceConnectionState.RTCIceConnectionStateFailed) {
        print('❌ ICE connection failed - check network settings');
      }
    };

    // Handle ICE gathering state changes
    _peerConnection!.onIceGatheringState = (RTCIceGatheringState state) {
      print('ICE Gathering State: $state');
    };

    // Handle data channel from Mac (Mac creates the channel)
    _peerConnection!.onDataChannel = (RTCDataChannel channel) {
      print('📬 Received data channel from Mac');
      _dataChannel = channel;
      _setupDataChannel();
    };

    print('✅ Peer connection created');
  }

  /// Convert connection state to string
  String _connectionStateToString(RTCPeerConnectionState state) {
    switch (state) {
      case RTCPeerConnectionState.RTCPeerConnectionStateNew:
        return 'new';
      case RTCPeerConnectionState.RTCPeerConnectionStateConnecting:
        return 'connecting';
      case RTCPeerConnectionState.RTCPeerConnectionStateConnected:
        return 'connected';
      case RTCPeerConnectionState.RTCPeerConnectionStateDisconnected:
        return 'disconnected';
      case RTCPeerConnectionState.RTCPeerConnectionStateFailed:
        return 'failed';
      case RTCPeerConnectionState.RTCPeerConnectionStateClosed:
        return 'closed';
    }
  }

  /// Setup data channel event handlers
  void _setupDataChannel() {
    if (_dataChannel == null) return;

    _dataChannel!.onDataChannelState = (RTCDataChannelState state) {
      switch (state) {
        case RTCDataChannelState.RTCDataChannelOpen:
          print('✅ WebRTC data channel opened');
          _isConnected = true;
          break;
        case RTCDataChannelState.RTCDataChannelClosed:
          print('⚠️  WebRTC data channel closed');
          _isConnected = false;
          break;
        default:
          break;
      }
    };

    _dataChannel!.onMessage = (RTCDataChannelMessage message) {
      try {
        final data = jsonDecode(message.text) as Map<String, dynamic>;
        _messageHandler?.call(data);
      } catch (error) {
        print('❌ Error parsing WebRTC message: $error');
      }
    };
  }

  /// Send a message via the data channel
  bool sendMessage(String type, dynamic payload) {
    if (_isConnected &&
        _dataChannel != null &&
        _dataChannel!.state == RTCDataChannelState.RTCDataChannelOpen) {
      try {
        final message = jsonEncode({'type': type, 'payload': payload});
        _dataChannel!.send(RTCDataChannelMessage(message));
        return true;
      } catch (error) {
        print('❌ Failed to send via WebRTC: $error');
        return false;
      }
    }
    return false;
  }

  /// Set message handler
  void onMessage(WebRTCMessageHandler handler) {
    _messageHandler = handler;
  }

  /// Set connection state change handler
  void onStateChange(ConnectionStateHandler handler) {
    _onConnectionStateChange = handler;
  }

  /// Check if WebRTC is connected
  bool get isWebRTCConnected => _isConnected;

  /// Cleanup resources
  void cleanup() {
    _dataChannel?.close();
    _dataChannel = null;

    _peerConnection?.close();
    _peerConnection = null;

    _isConnected = false;
  }
}

File: mobile/src/services/WebRTCService.ts
import {Socket} from 'socket.io-client';
import {RTCPeerConnection, RTCIceCandidate, RTCSessionDescription} from 'react-native-webrtc';

// Import types from the library
import type RTCDataChannel from 'react-native-webrtc/lib/typescript/RTCDataChannel';
import type RTCDataChannelEvent from 'react-native-webrtc/lib/typescript/RTCDataChannelEvent';
import type RTCIceCandidateEvent from 'react-native-webrtc/lib/typescript/RTCIceCandidateEvent';

export type WebRTCMessageHandler = (data: unknown) => void;

export type WebRTCNamespace = 'terminal' | 'code';

interface WebRTCMessage {
  namespace: string;
  action: string;
  payload: unknown;
  type?: string;
}

export class WebRTCService {
  private peerConnection: RTCPeerConnection | null = null;
  private dataChannel: RTCDataChannel | null = null;
  private socket: Socket;
  private isConnected: boolean = false;
  private messageHandler: WebRTCMessageHandler | null = null;
  private onConnectionStateChange: ((state: string) => void) | null = null;

  constructor(socket: Socket) {
    this.socket = socket;
    this.setupSignalingHandlers();
  }

  private setupSignalingHandlers() {
    // Handle WebRTC offer from Mac
    this.socket.on('webrtc:offer', async ({offer}) => {
      console.log('📡 Received WebRTC offer from Mac');
      await this.handleOffer(offer);
    });

    // Handle ICE candidates from Mac
    this.socket.on('webrtc:ice-candidate', async ({candidate}) => {
      console.log('🧊 Received ICE candidate from Mac');
      await this.handleIceCandidate(candidate);
    });
  }

  private async handleOffer(offer: {sdp: string; type: string}) {
    try {
      // Create peer connection if it doesn't exist
      if (!this.peerConnection) {
        await this.createPeerConnection();
      }

      // Set remote description (the offer from Mac)
      await this.peerConnection!.setRemoteDescription(
        new RTCSessionDescription({
          sdp: offer.sdp,
          type: offer.type as 'offer' | 'answer' | 'pranswer' | 'rollback',
        }),
      );

      console.log('✅ Remote description set');

      // Create answer
      const answer = await this.peerConnection!.createAnswer();

      // Set local description once
      await this.peerConnection!.setLocalDescription(answer);

      // Wait for ICE gathering to complete before sending answer (with timeout)
      await Promise.race([
        new Promise<void>(resolve => {
          if (this.peerConnection!.iceGatheringState === 'complete') {
            resolve();
          } else {
            const checkGathering = () => {
              console.log(`ICE Gathering State: ${this.peerConnection!.iceGatheringState}`);
              if (this.peerConnection!.iceGatheringState === 'complete') {
                resolve();
              }
            };
            this.peerConnection!.addEventListener('icegatheringstatechange', checkGathering, {
              once: false,
            });
            // Also check immediately in case it completed while we were setting up
            checkGathering();
          }
        }),
        new Promise<void>(resolve =>
          setTimeout(() => {
            console.log('⏱️  ICE gathering timeout - proceeding with available candidates');
            resolve();
          }, 3000),
        ),
      ]);

      console.log('📡 Sending WebRTC answer to Mac (with all ICE candidates)');

      // Send answer to Mac via signaling server (localDescription already set above)
      this.socket.emit('webrtc:answer', {
        answer: {
          sdp: this.peerConnection!.localDescription!.sdp,
          type: this.peerConnection!.localDescription!.type,
        },
      });
    } catch (error) {
      console.error('❌ Failed to handle offer:', error);
    }
  }

  private async handleIceCandidate(candidate: {
    candidate: string;
    sdpMid?: string;
    sdpMLineIndex?: number;
  }) {
    try {
      if (this.peerConnection && candidate.candidate) {
        await this.peerConnection.addIceCandidate(
          new RTCIceCandidate({
            candidate: candidate.candidate,
            sdpMid: candidate.sdpMid,
            sdpMLineIndex: candidate.sdpMLineIndex,
          }),
        );
        console.log('✅ ICE candidate added');
      }
    } catch (error) {
      console.error('❌ Failed to add ICE candidate:', error);
    }
  }

  private async createPeerConnection() {
    console.log('🔗 Creating WebRTC peer connection...');

    // Configure for local connections (no STUN needed for same network)
    const configuration = {
      iceServers: [{urls: 'stun:stun.l.google.com:19302'}], // Add STUN server to help discovery
      iceTransportPolicy: 'all' as 'all', // Allow both relay and host candidates
      iceCandidatePoolSize: 10,
    };

    this.peerConnection = new RTCPeerConnection(configuration);

    // Handle ICE candidates
    this.peerConnection.addEventListener(
      'icecandidate',
      (event: RTCIceCandidateEvent<'icecandidate'>) => {
        if (event.candidate) {
          console.log('🧊 Generated ICE candidate, sending to Mac');
          const candidateStr = event.candidate.candidate || '';
          const candidateType = candidateStr.split(' ')[7] || 'unknown';
          console.log(`   Type: ${candidateType}, Candidate: ${candidateStr.substring(0, 50)}...`);
          this.socket.emit('webrtc:ice-candidate', {
            candidate: {
              candidate: event.candidate.candidate,
              sdpMid: event.candidate.sdpMid,
              sdpMLineIndex: event.candidate.sdpMLineIndex,
            },
          });
        }
      },
    );

    // Handle connection state changes
    this.peerConnection.addEventListener('connectionstatechange', () => {
      const state = this.peerConnection?.connectionState || 'unknown';
      console.log(`WebRTC Connection State: ${state}`);

      if (state === 'connected') {
        this.isConnected = true;
        console.log('🎉 WebRTC P2P connection established!');
      } else if (state === 'disconnected' || state === 'failed' || state === 'closed') {
        this.isConnected = false;
        console.log('❌ WebRTC connection lost');
      }

      if (this.onConnectionStateChange) {
        this.onConnectionStateChange(state);
      }
    });

    // Handle ICE connection state changes (more detailed)
    this.peerConnection.addEventListener('iceconnectionstatechange', () => {
      const iceState = this.peerConnection?.iceConnectionState || 'unknown';
      console.log(`ICE Connection State: ${iceState}`);

      if (iceState === 'failed') {
        console.log('❌ ICE connection failed - check network settings');
      }
    });

    this.peerConnection.addEventListener('icegatheringstatechange', () => {
      const gatheringState = this.peerConnection?.iceGatheringState || 'unknown';
      console.log(`ICE Gathering State: ${gatheringState}`);
    });

    // Handle data channel from Mac (Mac creates the channel)
    this.peerConnection.addEventListener(
      'datachannel',
      (event: RTCDataChannelEvent<'datachannel'>) => {
        console.log('📬 Received data channel from Mac');
        this.dataChannel = event.channel;
        this.setupDataChannel();
      },
    );

    console.log('✅ Peer connection created');
  }

  private setupDataChannel() {
    if (!this.dataChannel) {
      return;
    }

    this.dataChannel.addEventListener('open', () => {
      console.log('✅ WebRTC data channel opened');
      this.isConnected = true;
    });

    this.dataChannel.addEventListener('close', () => {
      console.log('⚠️  WebRTC data channel closed');
      this.isConnected = false;
    });

    this.dataChannel.addEventListener('error', (error: Event) => {
      console.error('❌ WebRTC data channel error:', error);
      this.isConnected = false;
    });

    this.dataChannel.addEventListener('message', (event: MessageEvent) => {
      try {
        const data = JSON.parse(event.data as string);
        // Log all incoming messages for debugging
        console.log(`📨 WebRTC raw message: type=${data.type}`);
        if (this.messageHandler) {
          this.messageHandler(data);
        } else {
          console.warn('⚠️ WebRTC message received but no handler set!');
        }
      } catch (error) {
        console.error(
          '❌ Error parsing WebRTC message:',
          error,
          'Raw:',
          event.data?.substring?.(0, 100),
        );
      }
    });
  }

  /**
   * Send message via WebRTC data channel
   * Supports both namespace format (3 args) and legacy format (2 args)
   */
  public sendMessage(
    namespaceOrType: WebRTCNamespace | string,
    actionOrPayload: string | unknown,
    payload?: unknown,
  ): boolean {
    if (this.isConnected && this.dataChannel && this.dataChannel.readyState === 'open') {
      try {
        let message: WebRTCMessage;

        // Check if using new namespace format (3 arguments)
        if (payload !== undefined && typeof actionOrPayload === 'string') {
          // New format: namespace, action, payload
          message = {
            namespace: namespaceOrType,
            action: actionOrPayload,
            payload: payload,
          };
        } else {
          // Legacy format: type, payload
          // Convert to new format with 'terminal' namespace for backward compatibility
          message = {
            namespace: 'terminal',
            action: namespaceOrType,
            payload: actionOrPayload,
            // Keep old format for backward compatibility
            type: namespaceOrType,
          };
        }

        this.dataChannel.send(JSON.stringify(message));
        return true;
      } catch (error) {
        console.error('❌ Failed to send via WebRTC:', error);
        return false;
      }
    }
    return false;
  }

  public onMessage(handler: WebRTCMessageHandler) {
    this.messageHandler = handler;
  }

  public onStateChange(handler: (state: string) => void) {
    this.onConnectionStateChange = handler;
  }

  public isWebRTCConnected(): boolean {
    return this.isConnected;
  }

  public cleanup() {
    if (this.dataChannel) {
      this.dataChannel.close();
      this.dataChannel = null;
    }

    if (this.peerConnection) {
      this.peerConnection.close();
      this.peerConnection = null;
    }

    this.isConnected = false;
  }
}
